Tests/MongoSwiftSyncTests/UnifiedTestRunner/Matching.swift

```diff
@@ -361,6 +363,208 @@ func matchesEvents(expected: [ExpectedEvent], actual: [CommandEvent], entities:
     }
 }
 
+/// Test protocol used to indicate an error has one or more error codes and codenames. We use an array since
+/// BulkWriteErrors may have multiple failures and we need to be able to check if any of them match.
+protocol HasErrorCodes: MongoErrorProtocol {
+    var errorCodes: [MongoError.ServerErrorCode] { get }
+    var errorCodeNames: [String] { get }
+}
+
+extension MongoError.CommandError: HasErrorCodes {
+    var errorCodes: [MongoError.ServerErrorCode] { [self.code] }
+    var errorCodeNames: [String] { [self.codeName] }
+}
+
+extension MongoError.WriteError: HasErrorCodes {
+    var errorCodes: [MongoError.ServerErrorCode] {
+        if let code = self.writeFailure?.code {
+            return [code]
+        } else if let code = self.writeConcernFailure?.code {
+            return [code]
+        }
+        return []
+    }
+
+    var errorCodeNames: [String] {
+        if let codeName = self.writeFailure?.codeName {
+            return [codeName]
+        } else if let codeName = self.writeConcernFailure?.codeName {
+            return [codeName]
+        }
+        return []
+    }
+}
+
+extension MongoError.BulkWriteError: HasErrorCodes {
+    var errorCodes: [MongoError.ServerErrorCode] {
+        var codes = self.writeFailures?.map { $0.code } ?? []
+        if let wcCode = self.writeConcernFailure?.code {
+            codes.append(wcCode)
+        }
+        return codes
+    }
+
+    var errorCodeNames: [String] {
+        var codeNames = self.writeFailures?.map { $0.codeName } ?? []
+        if let wcCodeName = self.writeConcernFailure?.codeName {
+            codeNames.append(wcCodeName)
+        }
+        return codeNames
+    }
+}
+
+extension MongoErrorProtocol {
```

Comment:
I ended up just making this a method on `MongoErrorProtocol` instead; we only need to match on our own errors and it removes the need to do some casting